[gen_l10n] Warn users when placeholder types are converted to 'num' when using pluralization

thkim1011 opened the pull request:
See title. This PR refactors logging so that the logger is saved as an instance variable of the `LocalizationsGenerator` class, and adds a warning when there is a pluralization placeholder whose type isn't num.

Fixes #98868

Tested locally:
```
Multiple devices found:
macOS (desktop) • macos  • darwin-arm64   • macOS 12.4 21F79 darwin-arm (Rosetta)
Chrome (web)    • chrome • web-javascript • Google Chrome 103.0.5060.114
[1]: macOS (macos)
[2]: Chrome (chrome)
Please choose one (To quit, press "q/Q"): 2
Placeholders for plurals are automatically converted to type 'num'.
Placeholders for plurals are automatically converted to type 'num'.
Running "flutter pub get" in i18n_test...                          698ms
```
I didn't think I needed to add tests here since there's really no change other than logging.

thkim1011 committed:
init

HansMuller reviewed (commented):
You can ask @Hixie for a "test exemption" however it might be simplest to just include one.

thkim1011 committed:
asdf

thkim1011 committed:
fix tests

thkim1011 committed:
change warning slightly

thkim1011 commented:
Hey @HansMuller, I just made the change you recommended, added a test (turns out testing logs are not too bad), and fixed up the rest of the tests that were affected by changing the `LocalizationsGenerator` constructor.

thkim1011 commented on packages/flutter_tools/lib/src/localizations/gen_l10n.dart:
```diff
@@ -1141,6 +1148,16 @@ class LocalizationsGenerator {
+      if (message.isPlural) {
+        final Placeholder countPlaceholder = message.getCountPlaceholder();
+        if (countPlaceholder.type != null && countPlaceholder.type != 'num') {
+          logger.printWarning("Placeholders for plurals are automatically converted to type 'num'.");
```
Fixed!

thkim1011 committed:
fix test again

HansMuller commented on packages/flutter_tools/lib/src/localizations/gen_l10n.dart:
```diff
@@ -1141,6 +1148,22 @@ class LocalizationsGenerator {
       );
     });
 
+    // Warn users when placeholder types are overrided to 'num'.
```
This comment probably isn't necessary; at the moment it doesn't describe the entire block.

HansMuller commented on packages/flutter_tools/lib/src/localizations/gen_l10n.dart:
```diff
@@ -1141,6 +1148,22 @@ class LocalizationsGenerator {
 
+    // Warn users when placeholder types are overrided to 'num'.
+    for (final Message message in messages) {
+      if (message.isPlural) {
```
These tests/warnings are great.

HansMuller reviewed (approved):
The bots noticed some trailing spaces. Other than that and a trivial point about a comment, this is ready to land.

thkim1011 committed:
fix spacing and comment

thkim1011 committed:
fix lint
